lib/iris/analysis/stats.py

```diff
@@ -22,108 +22,45 @@
     """
+    Calculates the n-D Pearson's r correlation cube over the dimensions
+    associated with the given coordinates.
 
-    # Following lists to be filled with:
-    # indices of dimension we are not comparing
-    res_ind = []
-    # indices of dimensions we are comparing
-    slice_ind = []
-    for i, c in enumerate(cube_a.dim_coords):
-        if not c.name() in corr_coords:
-            res_ind.append(i)
-        else:
-            slice_ind.append(i)
-
-    # sanitise input
-    dim_coord_names = [c.name() for c in cube_a.dim_coords]
-    if corr_coords is None:
-        corr_coords = dim_coord_names
-
-    if ([c.name() for c in cube_a.dim_coords] !=
-            [c.name() for c in cube_b.dim_coords]):
-        raise ValueError("Cubes are incompatible.")
-
-    for c in corr_coords:
-        if c not in dim_coord_names:
-            raise ValueError("%s coord "
-                             "does not exist in cube." % c)
-
-    # Reshape data to be data to correlate in 0th dim and
-    # other grid points in 1st dim.
-    # Transpose to group the correlation data dims before the
-    # grid point dims.
-    data_a = cube_a.data.view()
-    data_b = cube_b.data.view()
-    dim_i_len = np.prod(np.array(cube_a.shape)[slice_ind])
-    dim_j_len = np.prod(np.array(cube_a.shape)[res_ind])
-    reshaped_a = data_a.transpose(slice_ind+res_ind)\
-                       .reshape(dim_i_len, dim_j_len)
-    reshaped_b = data_b.transpose(slice_ind+res_ind)\
-                       .reshape(dim_i_len, dim_j_len)
-
-    return reshaped_a, reshaped_b, res_ind
-
-
-def pearsonr(cube_a, cube_b, corr_coords=None):
-    """
-    Calculates the n-D Pearson's r correlation
-    cube over the dimensions associated with the
-    given coordinates.
-
-    Returns a cube of the correlation between the two
-    cubes along the dimensions of the given
-    coordinates, at each point in the remaining
+    Returns a cube of the correlation between the two cubes along the
+    dimensions of the given coordinates, at each point in the remaining
     dimensions of the cubes.
 
-    For example providing two time/altitude/latitude/longitude
-    cubes and corr_coords of 'latitude' and 'longitude' will result
-    in a time/altitude cube describing the latitude/longitude
-    (i.e. pattern) correlation at each time/altitude point.
+    For example providing two time/altitude/latitude/longitude cubes and
+    corr_coords of 'latitude' and 'longitude' will result in a time/altitude
+    cube describing the latitude/longitude (i.e. pattern) correlation at each
+    time/altitude point.
 
     Args:
 
     * cube_a, cube_b (cubes):
-        Between which the correlation field will be calculated.
-        Cubes should be the same shape and have the
-        same dimension coordinates.
+        Between which the correlation field will be calculated.  Cubes should
+        be the same shape and have the same dimension coordinates or cube_b
```

Comment:
Can we just find the smaller one and swap the inputs if necessary? If you make a temporary assignment it just creates a reference so doing something like:

```python
if cube_a.ndim < cube_b.ndim:
    a = cube_b
    b = cube_a
else:
    a = cube_a
    b = cube_b
```

(don't care about the naming, just an example) and then use `a` and `b` instead of `cube_a` and `cube_b` won't actually create copies of `cube_a` and `cube_b` so there are no extra memory issues to account for.